mayastor/src/bdev/nexus/nexus_label.rs

```diff
@@ -126,340 +141,221 @@ pub enum LabelError {
 
-        Ok(LabelData {
-            offset: label.secondary.lba_table * block_size,
-            buf,
-        })
-    }
+        partitions.push(GptEntry {
+            ent_type: GptGuid::from_str(Nexus::METADATA_PARTITION_TYPE_ID)
+                .unwrap(),
+            ent_guid: config.meta_guid,
+            ent_start: header.lba_start,
+            ent_end: data - 1,
+            ent_attr: 0,
+            ent_name: "MayaMeta".into(),
+        });
 
-    pub async fn write_labels(
-        &self,
-        target: &NexusLabel,
-        list: &[NexusChildLabel<'_>],
-    ) -> Result<(), LabelError> {
-        let primary = self.get_primary_data(target)?;
-        let secondary = self.get_secondary_data(target)?;
+        partitions.push(GptEntry {
+            ent_type: GptGuid::from_str(Nexus::METADATA_PARTITION_TYPE_ID)
+                .unwrap(),
+            ent_guid: config.data_guid,
+            ent_start: data,
+            ent_end: min(data + data_blocks - 1, header.lba_end),
+            ent_attr: 0,
+            ent_name: "MayaData".into(),
+        });
 
-        let mut futures = Vec::new();
+        Ok(partitions)
+    }
 
-        for label in list {
-            match label.get_label_status() {
-                NexusLabelStatus::Both => {
-                    // Nothing to do as both labels are already valid.
-                }
-                NexusLabelStatus::Primary => {
-                    // Only write out secondary as primary is already valid.
-                    futures.push(
-                        label.child.write_at(secondary.offset, &secondary.buf),
-                    );
-                }
-                NexusLabelStatus::Secondary => {
-                    // Only write out primary as secondary is already valid.
-                    futures.push(
-                        label.child.write_at(primary.offset, &primary.buf),
-                    );
+    // check if all children have valid labels
+    pub(crate) async fn validate_labels(&self) -> Result<(), LabelError> {
+        for child in self.children.iter() {
+            match child.probe_label().await {
+                Ok(label) => {
+                    if label.status != NexusLabelStatus::Both {
+                        return Err(LabelError::Generic {
+                            message: format!(
+                                "invalid label for child {}",
+                                child.name
+                            ),
+                        });
+                    }
                 }
-                NexusLabelStatus::Neither => {
-                    futures.push(
-                        label.child.write_at(primary.offset, &primary.buf),
-                    );
-                    futures.push(
-                        label.child.write_at(secondary.offset, &secondary.buf),
-                    );
+                Err(error) => {
+                    return Err(LabelError::Generic {
```

Comment:
Since this one has a source, I wonder if a `GenericSourced` error variant with a `source` would be more appropriate?

Reply:
In this case the "source" is a string rather than an error. snafu treats a "source" field automagically and so using that term here would only add to the confusion.

Reply:
Now removed entirely